src/backend/cdb/cdbmutate.c

```diff
@@ -1113,6 +1116,375 @@ request_explicit_motion(Plan *plan, Index resultRelationsIdx, List *rtable)
+		relation_close(relation, NoLock);
+	}
+}
+
+/*
+ * In legacy planner, we add a SplitUpdate node at top so that updating on distribution
+ * columns could be handled. The SplitUpdate will split each update into delete + insert.
+ */
+SplitUpdate*
+make_splitupdate(PlannerInfo *root, Plan *mt, Plan *subplan, RangeTblEntry *rte, Index resultRelationsIdx)
+{
+	AttrNumber		ctidColIdx = 0;
+	ListCell		*lct;
+	ListCell		*currAppendCell = NULL;
+	List			*deleteColIdx = NIL;
+	List			*insertColIdx = NIL;
+	List			*varsAbsent = NIL;
+	int				actionColIdx;
+	int				oidColIdx = 0;
+	int				attrIdx;
+	int				appendStart;
+	List			*splitUpdateTargetList = NIL;
+	Var				*var;
+	TargetEntry		*newTargetEntry;
+	SplitUpdate		*splitupdate;
+	DMLActionExpr	*actionExpr;
+	Relation		resultRelation;
+	TupleDesc		resultDesc;
+	bool			hasOids = false;
+
+	Assert(IsA(mt, ModifyTable));
+
+	/* Suppose we already hold locks before caller */
+	resultRelation = relation_open(rte->relid, NoLock);
+
+	failIfUpdateTriggers(resultRelation);
+
+	resultDesc = RelationGetDescr(resultRelation);
+
+	for (attrIdx = 1; attrIdx <= resultDesc->natts; ++attrIdx)
+	{
+		TargetEntry			*tle;
+		Var					*splitVar;
+		TargetEntry			*splitTargetEntry;
+		Form_pg_attribute	attr;
+
+		insertColIdx = lappend_int(insertColIdx, attrIdx);
+		deleteColIdx = lappend_int(deleteColIdx, attrIdx);
+
+		tle = (TargetEntry *) list_nth(subplan->targetlist, attrIdx - 1);
+
+		Assert(tle);
+
+		attr = resultDesc->attrs[attrIdx - 1];
+		if (attr->attisdropped)
+		{
+			Assert(IsA(tle->expr, Const) && ((Const *) tle->expr)->constisnull);
+
+			splitTargetEntry = makeTargetEntry((Expr *) copyObject(tle->expr), tle->resno, tle->resname, tle->resjunk);
+			splitUpdateTargetList = lappend(splitUpdateTargetList, splitTargetEntry);
+			continue;
+		}
+		else
+		{
+			splitVar = makeVar(OUTER, attrIdx, exprType((Node *) tle->expr),
+							   exprTypmod((Node *) tle->expr), 0 /* varlevelsup */);
+			splitTargetEntry = makeTargetEntry((Expr *) splitVar, tle->resno, tle->resname, tle->resjunk);
+			splitUpdateTargetList = lappend(splitUpdateTargetList, splitTargetEntry);
+		}
+
+		if (IsA(tle->expr, Var) &&
+			((Var *) tle->expr)->varnoold == resultRelationsIdx &&
+			((Var *) tle->expr)->varoattno == attrIdx)
+			continue;
+
+		varsAbsent = lappend(varsAbsent,
+							 makeVar(resultRelationsIdx, attrIdx, exprType((Node *) tle->expr),
+									 exprTypmod((Node *) tle->expr), 0 /* varlevelsup */));
+	}
+
+	if (resultRelation->rd_rel->relhasoids)
+		hasOids = true;
+
+
+	relation_close(resultRelation, NoLock);
+
+	Assert(attrIdx <= list_length(subplan->targetlist));
+	Assert(IsA(((TargetEntry *) list_nth(subplan->targetlist, attrIdx - 1))->expr, Var));
+
+	var = (Var *) ((TargetEntry *) list_nth(subplan->targetlist, attrIdx - 1))->expr;
+
+	/* Ctid should follow after normal attributes */
+	Assert((var->varno == resultRelationsIdx &&
```

Comment:
Hmm. I'm not sure that's a reliable assumption. ModifyTable node itself uses ExecFindJunkAttribute() to find it. I'd suggest doing the same here.